File: mem_probe/scene_discovery.py
"""Value-anchor scene_manager / SceneInfo klass discovery.

Algorithm (no IL2CPP base address needed):

  1. Capture (scene_uuid, dungeon_id) from TCP SyncDungeonData. SceneUuid is
     int64 and unique per dungeon instance — strong anchor.
  2. Cython AVX2 scan i64 == scene_uuid across heap. Few hits (typically <50).
  3. For each hit, try a small set of UID/scene_uuid field offsets to derive
     obj_base. Read 0x100-byte body once.
  4. Verify klass_ptr at obj_base is in GameAssembly.dll. Then scan body for
     dungeon_id (if available) and i32-aligned candidates as scene_id /
     layer.
  5. Use multi-field convergence (klass_ptr, scene_uuid_off, dungeon_id_off)
     — best signature wins.

Persists to anchors.json under `smart_locator.anchors.scene_manager`:
    obj_addr, klass_ptr, scene_uuid_off, dungeon_id_off, scene_id_off,
    layer_off, hp_width.

Note: with only one ground-truth scene_uuid per session this discovery has
weaker disambiguation than monster discovery (which converges across N
UUIDs). We rely on klass_ptr + dungeon_id co-presence.
"""
from __future__ import annotations

import bisect
import json
import os
import sys
import threading
import time
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from typing import Callable, Dict, List, Optional, Tuple

_HERE = os.path.dirname(os.path.abspath(__file__))
_SAO_AUTO = os.path.dirname(_HERE)
if _SAO_AUTO not in sys.path:
    sys.path.insert(0, _SAO_AUTO)

from mem_probe import cy_memscan as _cy

DEFAULT_BODY_BYTES = 0x180
DEFAULT_UID_OFF_TRIES = (0x10, 0x18, 0x20, 0x28, 0x30, 0x38, 0x40, 0x48, 0x50)
MAX_REGION_SIZE = 256 * 1024 * 1024


@dataclass
class SceneTruth:
    scene_uuid: int          # int64, unique per dungeon instance — primary anchor
    dungeon_id: int = 0      # int32, secondary — used for body validation
    scene_id: int = 0        # int32, optional — usually map id (e.g. 100, 200)
    layer: int = 0           # int32, optional — sub-area within scene
    difficulty: int = 0      # int32, optional

    def is_usable(self) -> bool:
        return self.scene_uuid > 0


def _hex(v: int) -> str:
    return f"0x{int(v):X}"


def _ga_module_range(pm) -> Tuple[int, int]:
    for m in pm.list_modules():
        if m.name.lower() == "gameassembly.dll":
            return m.base, m.base + m.size
    return 0, 0


def _scan_uuid_full_heap(pm, scene_uuid: int, *,
                         n_workers: int = 8) -> List[int]:
    """Cython AVX2 scan for i64 == scene_uuid. Returns absolute addresses."""
    needle = int(scene_uuid) & 0xFFFFFFFFFFFFFFFF
    regions = [r for r in pm.iter_regions(only_readable=True, only_private=True)
               if r.size <= MAX_REGION_SIZE]
    results: List[int] = []
    lock = threading.Lock()

    def scan_one(region):
        buf = pm.read_bytes(region.base, region.size)
        if buf is None:
            return
        offs = _cy.find_aligned_u64(buf, needle)
        if offs:
            with lock:
                for off in offs:
                    results.append(region.base + off)

    if n_workers > 1:
        with ThreadPoolExecutor(max_workers=n_workers) as ex:
            list(ex.map(scan_one, regions))
    else:
        for r in regions:
            scan_one(r)
    return results


def _find_i32_in_body(body: bytes, value: int) -> List[int]:
    """Return all 4-byte aligned offsets in body where i32 == value."""
    if value <= 0 or value > 0xFFFFFFFF:
        return []
    mask = _cy.narrow_u32_batch(body, int(value) & 0xFFFFFFFF)
    return [i * 4 for i, b in enumerate(mask) if b]


def discover_scene_klass(pm, truth: SceneTruth, *,
                         verbose: bool = True) -> Optional[dict]:
    """Find the SceneInfo / SceneManager klass + scene_uuid offset.

    Args:
        pm: StarProcess instance.
        truth: SceneTruth with at least scene_uuid set; dungeon_id is highly
               recommended for disambiguation.

    Returns:
        dict suitable for anchors.json scene_manager block, or None.
    """
    t0 = time.time()
    if not truth.is_usable():
        if verbose:
            print(f"[scene-discovery] truth missing scene_uuid")
        return None
    if verbose:
        print(f"[scene-discovery] anchor: scene_uuid={truth.scene_uuid} "
              f"dungeon_id={truth.dungeon_id} scene_id={truth.scene_id}")

    ga_base, ga_end = _ga_module_range(pm)
    if not ga_end and verbose:
        print(f"[scene-discovery] WARN: GameAssembly.dll not found")

    t_scan = time.time()
    hits = _scan_uuid_full_heap(pm, truth.scene_uuid, n_workers=8)
    if verbose:
        print(f"[scene-discovery] scene_uuid scan: {len(hits)} hits "
              f"({time.time()-t_scan:.2f}s, AVX2)")
    if not hits:
        return None

    # Per-candidate analysis: walk back, validate klass, find dungeon_id offset.
    # We collect all (klass_ptr, scene_uuid_off, dungeon_id_off) signatures
    # and pick the one that has a klass in GA AND (if available) finds
    # dungeon_id within the body.
    candidates: List[dict] = []
    for uid_addr in hits:
        for k1 in DEFAULT_UID_OFF_TRIES:
            obj_base = uid_addr - k1
            if obj_base < 0x10000:
                continue
            klass_blob = pm.read_bytes(obj_base, 8)
            if not klass_blob:
                continue
            klass_ptr = int.from_bytes(klass_blob, "little")
            if ga_end:
                if not (ga_base <= klass_ptr < ga_end):
                    continue
            body = pm.read_bytes(obj_base, DEFAULT_BODY_BYTES)
            if not body or len(body) < 0x40:
                continue

            # Try to find dungeon_id (i32) in body — narrows the right object
            dungeon_offs = _find_i32_in_body(body, truth.dungeon_id) if truth.dungeon_id else []
            scene_id_offs = _find_i32_in_body(body, truth.scene_id) if truth.scene_id else []
            difficulty_offs = _find_i32_in_body(body, truth.difficulty) if truth.difficulty else []
            layer_offs = _find_i32_in_body(body, truth.layer) if truth.layer else []

            # Score this candidate: how many ground-truth fields did we find?
            score = (
                (1 if dungeon_offs else 0)
                + (1 if scene_id_offs else 0)
                + (1 if difficulty_offs else 0)
                + (1 if layer_offs else 0)
            )

            candidates.append({
                "obj_addr": obj_base,
                "klass_ptr": klass_ptr,
                "scene_uuid_off": k1,
                "dungeon_id_offs": dungeon_offs,
                "scene_id_offs": scene_id_offs,
                "difficulty_offs": difficulty_offs,
                "layer_offs": layer_offs,
                "score": score,
            })

    if not candidates:
        if verbose:
            print(f"[scene-discovery] no candidate body had a GA-klass")
        return None

    # Pick the highest-scoring candidate. Tie-break: closest k1 to 0x10
    # (CharSerialize.SceneData layout; scene_uuid usually first field).
    candidates.sort(key=lambda c: (-c["score"], abs(c["scene_uuid_off"] - 0x10)))
    best = candidates[0]
    if best["score"] == 0 and len(candidates) > 1:
        if verbose:
            print(f"[scene-discovery] WARN: no body-side validation hits, "
                  f"picking highest-K1-quality candidate")

    klass_ptr = best["klass_ptr"]
    scene_uuid_off = best["scene_uuid_off"]
    obj_addr = best["obj_addr"]
    dungeon_id_off = (best["dungeon_id_offs"][0]
                      if best["dungeon_id_offs"] else -1)
    scene_id_off = (best["scene_id_offs"][0]
                    if best["scene_id_offs"] else -1)
    layer_off = (best["layer_offs"][0]
                 if best["layer_offs"] else -1)

    elapsed = time.time() - t0
    if verbose:
        print(f"[scene-discovery] WINNER (score={best['score']}/4):")
        print(f"   obj_addr        = 0x{obj_addr:X}")
        print(f"   klass_ptr       = 0x{klass_ptr:X}")
        print(f"   scene_uuid_off  = 0x{scene_uuid_off:X}")
        print(f"   dungeon_id_off  = {('0x%X' % dungeon_id_off) if dungeon_id_off >= 0 else 'unknown'}")
        print(f"   scene_id_off    = {('0x%X' % scene_id_off) if scene_id_off >= 0 else 'unknown'}")
        print(f"   layer_off       = {('0x%X' % layer_off) if layer_off >= 0 else 'unknown'}")
        print(f"   total cands     = {len(candidates)}")
        print(f"   elapsed         = {elapsed:.2f}s")

    return {
        "obj_addr": _hex(obj_addr),
        "klass_ptr": _hex(klass_ptr),
        "scene_uuid_off": int(scene_uuid_off),
        "dungeon_id_off": int(dungeon_id_off),
        "scene_id_off": int(scene_id_off),
        "layer_off": int(layer_off),
        "elapsed_s": elapsed,
        "score": int(best["score"]),
        "n_candidates": len(candidates),
    }


def persist_scene_anchor(anchors_path: str, payload: dict) -> None:
    """Write scene_manager block into anchors.json (schema v2)."""
    try:
        with open(anchors_path, "r", encoding="utf-8") as f:
            anchors = json.load(f)
    except (OSError, json.JSONDecodeError):
        anchors = {}
    sl = anchors.setdefault("smart_locator", {})
    nested = sl.setdefault("anchors", {})
    nested["scene_manager"] = payload
    sl["schema_version"] = max(2, int(sl.get("schema_version", 0) or 0))
    sl["scene_manager_set_at"] = time.time()
    with open(anchors_path, "w", encoding="utf-8") as f:
        json.dump(anchors, f, ensure_ascii=False, indent=2)


__all__ = [
    "SceneTruth",
    "discover_scene_klass",
    "persist_scene_anchor",
]

File: C#/src/SaoAuto.MemProbe/Discovery/SceneDiscovery.cs
using System.Buffers.Binary;

namespace SaoAuto.MemProbe.Discovery;

/// <summary>Ground-truth observation feeding <see cref="SceneDiscovery"/>.
/// Mirrors the Python <c>SceneTruth</c> dataclass.
/// <c>SceneUuid</c> is the only required field (int64, unique per dungeon
/// instance); the rest narrow the body match when present.</summary>
public sealed record SceneTruth(
    long SceneUuid,
    int DungeonId = 0,
    int SceneId = 0,
    int Layer = 0,
    int Difficulty = 0)
{
    public bool IsUsable => SceneUuid > 0;
}

/// <summary>Outcome of <see cref="SceneDiscovery.Discover"/>. Matches the
/// dict that Python <c>scene_discovery.discover_scene_klass</c> emits into
/// <c>anchors.json</c> under <c>smart_locator.anchors.scene_manager</c>.</summary>
public sealed record SceneDiscoveryResult(
    ulong ObjAddr,
    ulong KlassPtr,
    int SceneUuidOff,
    int DungeonIdOff,
    int SceneIdOff,
    int LayerOff,
    int Score,
    int CandidateCount);

/// <summary>
/// Port of <c>tools/mem_probe/scene_discovery.py</c>. Locates the
/// SceneInfo / SceneManager object purely from a TCP-derived
/// <c>(scene_uuid, dungeon_id)</c> pair — no IL2CPP base address needed.
///
/// Algorithm:
///   1. <see cref="MemoryScanner.ScanI64"/> the whole heap for
///      <c>scene_uuid</c> (8-byte aligned).
///   2. For each hit, walk a small set of candidate field offsets
///      (<see cref="DefaultUidOffTries"/>) backwards to a presumed object
///      base; read 8 bytes there as <c>klass_ptr</c>.
///   3. If a GameAssembly module range is supplied, drop candidates
///      whose <c>klass_ptr</c> is outside that range.
///   4. Read a <see cref="DefaultBodyBytes"/>-byte body once, scan it for
///      the optional ground-truth fields (dungeon_id / scene_id /
///      difficulty / layer) at any 4-byte aligned slot.
///   5. Score by how many of those secondary fields hit; tie-break by
///      proximity to <c>0x10</c> (CharSerialize.SceneData layout).
/// </summary>
public static class SceneDiscovery
{
    public const int DefaultBodyBytes = 0x180;
    public const int MaxRegionSize = 256 * 1024 * 1024;

    /// <summary>Default field offsets we probe backwards from a
    /// scene_uuid hit. Mirrors Python <c>DEFAULT_UID_OFF_TRIES</c>.</summary>
    public static readonly int[] DefaultUidOffTries =
        { 0x10, 0x18, 0x20, 0x28, 0x30, 0x38, 0x40, 0x48, 0x50 };

    /// <summary>Run the discovery pipeline.</summary>
    /// <param name="memory">Process memory source.</param>
    /// <param name="truth">Must have <see cref="SceneTruth.SceneUuid"/> &gt; 0.</param>
    /// <param name="gaBase">GameAssembly.dll base; 0 to skip klass_ptr range check.</param>
    /// <param name="gaEnd">GameAssembly.dll base + size; 0 to skip.</param>
    /// <param name="bodyBytes">How many bytes to read per candidate body.</param>
    /// <param name="uidOffTries">Override the candidate offset list.</param>
    /// <param name="rescanMaxHits">Cap on the heap scan.</param>
    /// <returns>Best-scoring candidate, or <c>null</c> if none survived.</returns>
    public static SceneDiscoveryResult? Discover(
        IMemorySource memory,
        SceneTruth truth,
        ulong gaBase = 0,
        ulong gaEnd = 0,
        int bodyBytes = DefaultBodyBytes,
        IReadOnlyList<int>? uidOffTries = null,
        int rescanMaxHits = MemoryScanner.DefaultMaxHits)
    {
        ArgumentNullException.ThrowIfNull(memory);
        ArgumentNullException.ThrowIfNull(truth);
        if (!truth.IsUsable) return null;

        var tries = uidOffTries ?? DefaultUidOffTries;
        var hits = MemoryScanner.ScanI64(
            memory, truth.SceneUuid,
            maxHits: rescanMaxHits,
            maxRegionSize: MaxRegionSize);
        if (hits.Count == 0) return null;

        var candidates = new List<Candidate>();
        foreach (var uidAddr in hits)
        {
            foreach (var k1 in tries)
            {
                if (uidAddr < (ulong)k1) continue;
                ulong objBase = uidAddr - (ulong)k1;
                if (objBase < 0x10000UL) continue;

                var klassBlob = memory.ReadBytes(objBase, 8);
                if (klassBlob is null) continue;
                ulong klassPtr = BinaryPrimitives.ReadUInt64LittleEndian(klassBlob);
                if (gaEnd != 0 && (klassPtr < gaBase || klassPtr >= gaEnd)) continue;

                var body = memory.ReadBytes(objBase, bodyBytes);
                if (body is null || body.Length < 0x40) continue;

                int dungeonOff = truth.DungeonId != 0 ? FirstI32Offset(body, truth.DungeonId) : -1;
                int sceneIdOff = truth.SceneId != 0 ? FirstI32Offset(body, truth.SceneId) : -1;
                int diffOff = truth.Difficulty != 0 ? FirstI32Offset(body, truth.Difficulty) : -1;
                int layerOff = truth.Layer != 0 ? FirstI32Offset(body, truth.Layer) : -1;

                int score = (dungeonOff >= 0 ? 1 : 0)
                    + (sceneIdOff >= 0 ? 1 : 0)
                    + (diffOff >= 0 ? 1 : 0)
                    + (layerOff >= 0 ? 1 : 0);

                candidates.Add(new Candidate(
                    objBase, klassPtr, k1,
                    dungeonOff, sceneIdOff, layerOff,
                    score));
            }
        }

        if (candidates.Count == 0) return null;

        candidates.Sort((a, b) =>
        {
            int byScore = b.Score.CompareTo(a.Score);
            if (byScore != 0) return byScore;
            return Math.Abs(a.SceneUuidOff - 0x10).CompareTo(Math.Abs(b.SceneUuidOff - 0x10));
        });
        var best = candidates[0];

        return new SceneDiscoveryResult(
            ObjAddr: best.ObjAddr,
            KlassPtr: best.KlassPtr,
            SceneUuidOff: best.SceneUuidOff,
            DungeonIdOff: best.DungeonIdOff,
            SceneIdOff: best.SceneIdOff,
            LayerOff: best.LayerOff,
            Score: best.Score,
            CandidateCount: candidates.Count);
    }

    /// <summary>Return the first 4-byte-aligned offset in
    /// <paramref name="body"/> where the i32 equals
    /// <paramref name="value"/>, or -1 if absent. Mirrors the Python
    /// helper that returned the full list and the caller picked [0].</summary>
    public static int FirstI32Offset(ReadOnlySpan<byte> body, int value)
    {
        Span<byte> needle = stackalloc byte[4];
        BinaryPrimitives.WriteInt32LittleEndian(needle, value);
        int last = body.Length - 4;
        for (int i = 0; i <= last; i += 4)
        {
            if (body[i] == needle[0] && body[i + 1] == needle[1]
                && body[i + 2] == needle[2] && body[i + 3] == needle[3])
            {
                return i;
            }
        }
        return -1;
    }

    private readonly record struct Candidate(
        ulong ObjAddr,
        ulong KlassPtr,
        int SceneUuidOff,
        int DungeonIdOff,
        int SceneIdOff,
        int LayerOff,
        int Score);
}
